### PR TITLE
Fix CGAS overflow bug

### PR DESCRIPTION
**Pull request by Dentosal**
I'm not sure if this is actually the correct approach, so please review carefully.

Also fortifies some gas arithmetics using `checked_*` methods, as per our discussion with @Voxelot.

### CODE REVIEW

File: src/interpreter/flow.rs [hunk 1]
```diff
@@ -158,6 +160,13 @@ where
         // credit contract asset_id balance
         self.balance_increase(call.to(), &asset_id, b)?;
 
+        let forward_gas_amount = cmp::min(self.registers[REG_GGAS], d);
+
+        // subtract gas
```
Comment (Voxelot) [inline src/interpreter/flow.rs:165]: this is the fix I had in mind 👍🏻

**Review by adlerjohn — COMMENTED** [review]
CI failing

**Comment by Dentosal** [comment]
This seems to introduce a new bug, caught by the `profile-gas` tests only. Not sure if the bug is in tests or code, looking at it now.

File: src/interpreter/gas.rs [hunk 2]
```diff
@@ -75,13 +75,19 @@ impl<S> Interpreter<S> {
         }
 
         if gas > self.registers[REG_CGAS] {
-            self.registers[REG_GGAS] -= self.registers[REG_CGAS];
+            self.registers[REG_GGAS] = self.registers[REG_CGAS]
+                .checked_sub(gas)
+                .ok_or(RuntimeError::halt_on_bug("gas invariant violation"))?;
```
Comment (Voxelot) [inline src/interpreter/gas.rs:80]:
```suggestion
            self.registers[REG_GGAS] = self.registers[REG_GGAS]
                .checked_sub(self.registers[REG_CGAS])
                .ok_or(RuntimeError::halt_on_bug("gas invariant violation"))?;
```

**Review by Voxelot — APPROVED** [review]
lgtm

**Pull request merged**